src/platforms/esp32/main/spi_driver.c

```diff
@@ -249,8 +264,17 @@ static void spidriver_consume_mailbox(Context *ctx)
             ret = ERROR_ATOM;
     }
 
-    mailbox_destroy_message(message);
+    dictionary_put(&ctx->dictionary, ctx, spi_driver, ret);
+    term ret_msg;
+    if (UNLIKELY(memory_ensure_free(ctx, 3) != MEMORY_GC_OK)) {
```

Comment:
wouldn't be easier to call `memory_ensure_free` first (at line 237)?

Reply:
Yes, but what happens when `memory_ensure_free` is called again on line 176, when allocating the tuple that eventually becomes `ret`?  For example, if there is only enough space in the heap for one tuple, then neither call will result in a GC, but the second alloc will walk over the stack.